server/routes/transferRouter.js

```diff
@@ -76,7 +77,18 @@ transferRouter.post(
         // TODO: get only transferrable tokens
         result = await walletLogin.transferBundle(walletSender, walletReceiver, req.body.bundle.bundle_size, claim);
       }
-      const transferService = new TransferService(session);
+
+      // send message
+      if (result.state === Transfer.STATE.completed) {
+        // just send message in production 
+        if(process.env.NODE_ENV !== "test"){
+          await transferService.sendMessage(result.id);
```

Comment:
One way we could make this a bit more robust is to raise the messages as a domain-event (a separate table that stores all the events raised) with a payload that holds the message and a status(that can be  `raised`, `sent`). 
The raised domain-event will be stored and should be in the same transaction along with the transfer logic.
After the first part is complete, we could proceed to `dispatch` the domain event payload (the message) to the queue and mark the domain-event as sent in a second db call that probably doesn't need to be in a transaction since it deals with just one table.

Reply:
Do we have some 3rd party libraries to implement this pattern? Or we should do a module about this by ourselves? I think this might be a common demand in different our project?

Reply:
You are right, we need this in whichever service we emit/receive messages and store/process them. I started a separate project (to be treated as a submodule) https://github.com/Greenstand/domain-events. I am not sure if there is a NPM library that can meet our needs exactly.  It might easier to use our own given we also need to store these events in a database.